File: 2014-12/marathon.py
# TLE, 12/15
import sys
sys.stdin = open("marathon.in")
sys.stdout = open("marathon.out", 'w')
from functools import lru_cache
sys.setrecursionlimit(3 * 10 ** 5)

n, k = map(int, input().split())
checkpoints = [tuple(map(int, input().split())) for _ in range(n)]

@lru_cache(None)
def dp(i, skips):
    if i >= n - 1:
        return 0
    ret = float('inf')
    # try all skip options
    for newSkips in range(1, k - skips + 1):
        newLoc = i + newSkips + 1
        if newLoc > n - 1:
            continue
        d = abs(checkpoints[newLoc][0] - checkpoints[i][0]) + abs(checkpoints[newLoc][1] - checkpoints[i][1])
        ret = min(ret, d + dp(newLoc, skips + newSkips))
    # take normal
    d = abs(checkpoints[i + 1][0] - checkpoints[i][0]) + abs(checkpoints[i + 1][1] - checkpoints[i][1])
    ret = min(ret, d + dp(i + 1, skips))
    return ret

ans = dp(0, 0)
print(ans)

'''
dp(i, skips) = min distance at checkpoint i using [skips] skips
'''

File: 2014-12/marathon.cpp
#include <iostream>
#include <vector>

using namespace std;

int n, k;
vector<pair<int,int>> checkpoints;
vector<vector<int>> memo;

int dp(int i, int skips) {
    if (i >= n - 1) {
        return 0;
    }
    if (memo[i][skips] != -1){
        return memo[i][skips];
    }
    int ret = 1000000;
    for (int newSkips = 1; newSkips < k - skips + 1; newSkips++) {
        int newLoc = i + newSkips + 1;
        if (newLoc > n - 1) {
            continue;
        }
        int d = abs(checkpoints[newLoc].first - checkpoints[i].first) + abs(checkpoints[newLoc].second - checkpoints[i].second);
        ret = min(ret, d + dp(newLoc, skips + newSkips));
    }
    int d = abs(checkpoints[i + 1].first - checkpoints[i].first) + abs(checkpoints[i + 1].second - checkpoints[i].second);
    ret = min(ret, d + dp(i + 1, skips));
    return memo[i][skips] = ret;
}

int main() {
    freopen("marathon.in", "r", stdin);
    freopen("marathon.out", "w", stdout);

    cin >> n >> k;
    checkpoints.resize(n);
    
    for (int i = 0; i < n; i++) {
        cin >> checkpoints[i].first >> checkpoints[i].second;
    }

    memo.assign(n, vector<int>(k + 1, -1));

    int ans = dp(0, 0);
    cout << ans;
}
